common/flatpak-dir.c

```diff
@@ -7548,6 +7567,112 @@ apply_extra_data (FlatpakDir   *self,
   return TRUE;
 }
 
+/* Check the user’s parental controls allow installation of @ref by looking at
+ * its cached @deploy_data, which contains its content rating as extracted from
+ * its AppData when it was originally downloaded. That’s compared to the
+ * parental controls policy loaded from the #MctManager.
+ *
+ * If @ref should not be installed, an error is returned. */
+static gboolean
+flatpak_dir_check_parental_controls (FlatpakDir    *self,
+                                     const char    *ref,
+                                     GVariant      *deploy_data,
+                                     GCancellable  *cancellable,
+                                     GError       **error)
+{
+#ifdef HAVE_LIBMALCONTENT
+  g_autoptr(GError) local_error = NULL;
+  const char *on_session = g_getenv ("FLATPAK_SYSTEM_HELPER_ON_SESSION");
+  g_autoptr(GDBusConnection) dbus_connection = NULL;
+  g_autoptr(MctManager) manager = NULL;
+  g_autoptr(MctAppFilter) app_filter = NULL;
+  const char *content_rating_type;
+  g_autoptr(GHashTable) content_rating = NULL;
+  g_autoptr(AutoPolkitAuthority) authority = NULL;
+  g_autoptr(AutoPolkitDetails) details = NULL;
+  g_autoptr(AutoPolkitSubject) subject = NULL;
+  g_autoptr(AutoPolkitAuthorizationResult) result = NULL;
+  gboolean authorized;
+
+  /* The ostree-metadata and appstream/ branches should not have any parental
+   * controls restrictions. */
+  if (!g_str_has_prefix (ref, "app/") && !g_str_has_prefix (ref, "runtime/"))
+    return TRUE;
+
+  g_debug ("Getting parental controls details for %s from %s",
+           ref, flatpak_deploy_data_get_origin (deploy_data));
+
+  dbus_connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, cancellable, &local_error);
+  if (dbus_connection == NULL)
+    {
+      if (on_session != NULL)
+        {
+          /* FIXME: Instead of skipping the parental controls check in the test
+           * environment, make a mock service for it.
+           * https://phabricator.endlessm.com/T25340 */
+          g_debug ("Skipping parental controls check for %s since the "
+                   "system bus is unavailable in the test environment", ref);
+          return TRUE;
```

Comment:
a) Yes, we want a mock thing for the tests
and
b) This check is very easy to disable, maybe we can make respecting on_session a configure option that only developers enable.

Reply:
> a) Yes, we want a mock thing for the tests

Can that be deferred to some follow-up work? I will start writing it this afternoon (it needs doing for various reasons), but it’s not going to be ready before next week.

> b) This check is very easy to disable, maybe we can make respecting on_session a configure option that only developers enable.

I can do that, but again I’m a bit hesitant to be testing code which is different from what goes out into production (even if the differences are small). I think a strategy for testing flatpak (as a whole, not just the parental controls stuff) by redirecting its system bus connection needs to be looked at, and that’s not something I’ve spent any real time on yet.

Could this also be deferred to follow up work? I don’t want to rush into clamping down the system bus connection only to end up making it a complete pain to test flatpak during development.

Reply:
Yeah, delaying both these is fine to me.

Reply:
Filed
 * https://github.com/flatpak/flatpak/issues/2993
 * https://github.com/flatpak/flatpak/issues/2994